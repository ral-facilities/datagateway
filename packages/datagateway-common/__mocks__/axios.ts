import axios from 'axios';

const requests = {
  get: vi.fn(() => Promise.resolve({ data: {} })),
  post: vi.fn(() => Promise.resolve({ data: {} })),
  delete: vi.fn(() => Promise.resolve({ data: {} })),
  CancelToken: axios.CancelToken,
  AxiosError: axios.AxiosError,
  isAxiosError: axios.isAxiosError,
};

export default requests;
