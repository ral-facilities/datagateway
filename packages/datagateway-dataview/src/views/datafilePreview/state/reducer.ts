import DATAFILE_PREVIEWER_DEFAULT from '../defaults';
import {
  DecrementDatafilePreviewerZoomLevelType,
  IncrementDatafilePreviewerZoomLevelType,
  ResetDatafilePreviewerZoomLevelType,
  ToggleDatafilePreviewerDetailsPanePayload,
  ToggleDatafilePreviewerDetailsPaneType,
} from './actions';
import { DGDataViewState } from '../../../state/app.types';

/**
 * Defines the shape of the redux state for {@link DatafilePreviewer}
 */
interface DatafilePreviewerState {
  /**
   * Whether the details pane should be shown or not.
   */
  isDetailsPaneShown: boolean;

  /**
   * The current zoom level of the previewer, in percentage. For example,
   * 100% maps to 100, 110% maps to 110, 80% maps to 80, etc.
   */
  zoomLevel: number;
}

function toggleDatafilePreviewerDetailsPane(
  state: DGDataViewState,
  payload: ToggleDatafilePreviewerDetailsPanePayload
): DGDataViewState {
  return {
    ...state,
    datafilePreviewer: {
      ...state.datafilePreviewer,
      isDetailsPaneShown: payload.shouldShow,
    },
  };
}

function incDatafilePreviewerZoomLevel(
  state: DGDataViewState
): DGDataViewState {
  return {
    ...state,
    datafilePreviewer: {
      ...state.datafilePreviewer,
      zoomLevel:
        state.datafilePreviewer.zoomLevel +
        DATAFILE_PREVIEWER_DEFAULT.zoomLevelStep,
    },
  };
}

function decDatafilePreviewerZoomLevel(
  state: DGDataViewState
): DGDataViewState {
  return {
    ...state,
    datafilePreviewer: {
      ...state.datafilePreviewer,
      // zoom level should not go lower than zoomLevelStep
      // otherwise will result in zoom level <= 0.
      zoomLevel: Math.max(
        DATAFILE_PREVIEWER_DEFAULT.minZoomLevel,
        state.datafilePreviewer.zoomLevel -
          DATAFILE_PREVIEWER_DEFAULT.zoomLevelStep
      ),
    },
  };
}

function resetDatafilePreviewerZoomLevel(
  state: DGDataViewState
): DGDataViewState {
  return {
    ...state,
    datafilePreviewer: {
      ...state.datafilePreviewer,
      zoomLevel: DATAFILE_PREVIEWER_DEFAULT.zoomLevel,
    },
  };
}

const datafilePreviewerInitialState: DatafilePreviewerState = {
  isDetailsPaneShown: DATAFILE_PREVIEWER_DEFAULT.showDetailsPane,
  zoomLevel: DATAFILE_PREVIEWER_DEFAULT.zoomLevel,
};

const datafilePreviewerReducer = {
  [ToggleDatafilePreviewerDetailsPaneType]: toggleDatafilePreviewerDetailsPane,
  [IncrementDatafilePreviewerZoomLevelType]: incDatafilePreviewerZoomLevel,
  [DecrementDatafilePreviewerZoomLevelType]: decDatafilePreviewerZoomLevel,
  [ResetDatafilePreviewerZoomLevelType]: resetDatafilePreviewerZoomLevel,
};

export {
  datafilePreviewerInitialState,
  datafilePreviewerReducer,
  toggleDatafilePreviewerDetailsPane,
  incDatafilePreviewerZoomLevel,
  decDatafilePreviewerZoomLevel,
  resetDatafilePreviewerZoomLevel,
};
export type { DatafilePreviewerState };
