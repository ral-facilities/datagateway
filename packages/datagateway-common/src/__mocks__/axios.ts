// TODO: move __mocks__ folder back to package root once facebook/create-react-app#7539 is fixed

import axios from 'axios';

const requests = {
  get: jest.fn(() => Promise.resolve({ data: {} })),
  post: jest.fn(() => Promise.resolve({ data: {} })),
  delete: jest.fn(() => Promise.resolve({ data: {} })),
  CancelToken: axios.CancelToken,
  AxiosError: axios.AxiosError,
};

export default requests;
