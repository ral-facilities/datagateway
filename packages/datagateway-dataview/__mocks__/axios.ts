import axios from 'axios';

const requests = {
  get: vi.fn((path) => {
    if (path === '/datagateway-dataview-settings.json') {
      return Promise.resolve({
        data: {
          facilityName: 'Generic',
          idsUrl: 'ids',
          apiUrl: 'api',
          downloadApiUrl: 'download',
          'ui-strings': '/res/default.json',
        },
      });
    } else {
      return Promise.resolve({
        data: {},
      });
    }
  }),
  post: vi.fn(() => Promise.resolve({ data: {} })),
  delete: vi.fn(() => Promise.resolve({ data: {} })),
  CancelToken: axios.CancelToken,
  isAxiosError: axios.isAxiosError,
};

export default requests;
