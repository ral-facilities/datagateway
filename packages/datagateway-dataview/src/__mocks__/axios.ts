// TODO: move __mocks__ folder back to package root once facebook/create-react-app#7539 is fixed

import axios from 'axios';

const requests = {
  get: jest.fn((path) => {
    if (path === '/datagateway-dataview-settings.json') {
      return Promise.resolve({
        data: {
          facilityName: 'Generic',
          idsUrl: 'ids',
          apiUrl: 'api',
          downloadApiUrl: 'download',
          'ui-strings': '/res/default.json',
        },
      });
    } else {
      return Promise.resolve({
        data: {},
      });
    }
  }),
  post: jest.fn(() => Promise.resolve({ data: {} })),
  delete: jest.fn(() => Promise.resolve({ data: {} })),
  CancelToken: axios.CancelToken,
  isAxiosError: axios.isAxiosError,
};

export default requests;
