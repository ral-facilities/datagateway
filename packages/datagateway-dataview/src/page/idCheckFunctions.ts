import axios, { AxiosResponse } from 'axios';
import {
  handleICATError,
  Dataset,
  Investigation,
  ConfigureURLsType,
  readSciGatewayToken,
} from 'datagateway-common';
import { Datafile } from 'datagateway-common/lib/app.types';
import { Middleware, Dispatch, AnyAction } from 'redux';
import memoize from 'lodash.memoize';

let apiUrl = '';

// this is so that idCheckFunctions have access to the apiUrl
export const saveApiUrlMiddleware: Middleware = (() =>
  (next: Dispatch<AnyAction>) =>
  (action: AnyAction): AnyAction => {
    if (action.type === ConfigureURLsType) {
      apiUrl = action.payload.urls.apiUrl;
    }

    return next(action);
  }) as Middleware;

const unmemoizedCheckInvestigationId = (
  investigationId: number,
  datasetId: number
): Promise<boolean> => {
  return axios
    .get(`${apiUrl}/datasets/findone`, {
      params: {
        where: {
          id: {
            eq: datasetId,
          },
        },
        include: '"investigation"',
      },
      headers: {
        Authorization: `Bearer ${readSciGatewayToken().sessionId}`,
      },
    })
    .then((response: AxiosResponse<Dataset>) => {
      return response.data.investigation?.id === investigationId;
    })
    .catch((error) => {
      handleICATError(error);
      return false;
    });
};

// we memoize so that we "remember" past promises so we don't redo to
// improves performance if a user revisits a specific view.
// However, depending on how users use the software, this
// may cause a memory leak since the cache size is infinite. In that case,
// we'd have to create our own memoize function with a cache limit
export const checkInvestigationId = memoize(
  unmemoizedCheckInvestigationId,
  (...args) => JSON.stringify(args)
);

const unmemoizedCheckInstrumentAndFacilityCycleId = (
  instrumentId: number,
  facilityCycleId: number,
  investigationId: number
): Promise<boolean> => {
  return axios
    .get(
      `${apiUrl}/instruments/${instrumentId}/facilitycycles/${facilityCycleId}/investigations/`,
      {
        params: {
          where: {
            id: {
              eq: investigationId,
            },
          },
        },
        headers: {
          Authorization: `Bearer ${readSciGatewayToken().sessionId}`,
        },
      }
    )
    .then((response: AxiosResponse<Investigation[]>) => {
      return response.data.length > 0;
    })
    .catch((error) => {
      handleICATError(error);
      return false;
    });
};

export const checkInstrumentAndFacilityCycleId = memoize(
  unmemoizedCheckInstrumentAndFacilityCycleId,
  (...args) => JSON.stringify(args)
);

const unmemoizedCheckStudyId = (
  studyId: number,
  investigationId: number
): Promise<boolean> => {
  const params = new URLSearchParams();
  params.append(
    'where',
    JSON.stringify({
      id: { eq: investigationId },
    })
  );
  params.append(
    'where',
    JSON.stringify({
      'studyInvestigations.study.id': {
        eq: studyId,
      },
    })
  );
  return axios
    .get(`${apiUrl}/investigations/`, {
      params,
      headers: {
        Authorization: `Bearer ${readSciGatewayToken().sessionId}`,
      },
    })
    .then((response: AxiosResponse<Investigation[]>) => {
      return response.data.length > 0;
    })
    .catch((error) => {
      handleICATError(error);
      return false;
    });
};

export const checkStudyId = memoize(unmemoizedCheckStudyId, (...args) =>
  JSON.stringify(args)
);

const unmemoizedCheckInstrumentId = (
  instrumentId: number,
  studyId: number
): Promise<boolean> => {
  const params = new URLSearchParams();
  params.append(
    'where',
    JSON.stringify({
      id: { eq: studyId },
    })
  );
  params.append(
    'where',
    JSON.stringify({
      'studyInvestigations.investigation.investigationInstruments.instrument.id':
        {
          eq: instrumentId,
        },
    })
  );
  return axios
    .get(`${apiUrl}/studies/`, {
      params,
      headers: {
        Authorization: `Bearer ${readSciGatewayToken().sessionId}`,
      },
    })
    .then((response: AxiosResponse<Investigation[]>) => {
      return response.data.length > 0;
    })
    .catch((error) => {
      handleICATError(error);
      return false;
    });
};

export const checkInstrumentId = memoize(
  unmemoizedCheckInstrumentId,
  (...args) => JSON.stringify(args)
);

export const unmemoizedCheckProposalName = (
  proposalName: string,
  investigationId: number
): Promise<boolean> => {
  return axios
    .get(`${apiUrl}/investigations/${investigationId}`, {
      headers: {
        Authorization: `Bearer ${readSciGatewayToken().sessionId}`,
      },
    })
    .then((response: AxiosResponse<Investigation>) => {
      return response.data.name === proposalName;
    })
    .catch((error) => {
      handleICATError(error);
      return false;
    });
};

export const checkProposalName = memoize(
  unmemoizedCheckProposalName,
  (...args) => JSON.stringify(args)
);

export const unmemoizedCheckDatafileId = (
  investigationId: Investigation['id'],
  datasetId: Dataset['id'],
  datafileId: Datafile['id']
): Promise<boolean> =>
  axios
    .get(`${apiUrl}/datafiles`, {
      params: {
        include: JSON.stringify(['dataset', 'dataset.investigation']),
        where: JSON.stringify({ id: { eq: datafileId } }),
      },
      headers: {
        Authorization: `Bearer ${readSciGatewayToken().sessionId}`,
      },
    })
    .then((response: AxiosResponse<Datafile[]>) => {
      if (response.data.length <= 0) return false;
      const datafile = response.data[0];
      return (
        datafile.id === datafileId &&
        datafile.dataset?.id === datasetId &&
        datafile.dataset?.investigation?.id === investigationId
      );
    })
    .catch((error) => {
      handleICATError(error);
      return false;
    });

export const checkDatafileId = memoize(unmemoizedCheckDatafileId, (...args) =>
  JSON.stringify(args)
);
