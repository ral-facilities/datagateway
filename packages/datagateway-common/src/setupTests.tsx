/* eslint-disable @typescript-eslint/no-empty-function */
import '@testing-library/jest-dom';
import React from 'react';
import { Action } from 'redux';
import { StateType } from './state/app.types';
import { initialState } from './state/reducers/dgcommon.reducer';
import { setLogger } from 'react-query';
import { WrapperComponent } from '@testing-library/react-hooks';
import { QueryClient, QueryClientProvider } from 'react-query';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { createMemoryHistory, History } from 'history';
import failOnConsole from 'jest-fail-on-console';

failOnConsole();

jest.setTimeout(20000);

if (typeof window.URL.createObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
  Object.defineProperty(window.URL, 'createObjectURL', {
    value: () => 'testObjectUrl',
  });
}

if (typeof window.URL.revokeObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
  Object.defineProperty(window.URL, 'revokeObjectURL', {
    value: () => {},
  });
}

// Add in ResizeObserver as it's not in Jest's environment
global.ResizeObserver = require('resize-observer-polyfill');

if (!global.structuredClone) {
  // structuredClone not available in jest/node <17, so this is a quick polyfill that should do the exact same thing
  global.structuredClone = (obj: unknown) => JSON.parse(JSON.stringify(obj));
}

// these are used for testing async actions
export let actions: Action[] = [];
export const resetActions = (): void => {
  actions = [];
};
export const getState = (): Partial<StateType> => ({ dgcommon: initialState });
export const dispatch = (action: Action): void | Promise<void> => {
  if (typeof action === 'function') {
    action(dispatch, getState);
    return Promise.resolve();
  } else {
    actions.push(action);
  }
};

// silence react-query errors
setLogger({
  log: console.log,
  warn: console.warn,
  error: jest.fn(),
});

// mock retry function to ensure it doesn't slow down query failure tests
jest.mock('./api/retryICATErrors', () => ({
  __esModule: true,
  default: jest.fn().mockReturnValue(false),
}));

export const createTestQueryClient = (): QueryClient =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });

export const createReactQueryWrapper = (
  history: History = createMemoryHistory()
): WrapperComponent<unknown> => {
  const testQueryClient = createTestQueryClient();
  const state = {
    dgcommon: {
      ...initialState,
      facilityName: 'TEST',
      urls: {
        apiUrl: 'https://example.com/api',
        icatUrl: 'https://example.com/icat',
        idsUrl: 'https://example.com/ids',
        downloadApiUrl: 'https://example.com/topcat',
      },
    },
  };

  const mockStore = configureStore([thunk]);

  const wrapper: WrapperComponent<unknown> = ({ children }) => (
    <Provider store={mockStore(state)}>
      <Router history={history}>
        <QueryClientProvider client={testQueryClient}>
          {children}
        </QueryClientProvider>
      </Router>
    </Provider>
  );
  return wrapper;
};

// MUI date pickers default to mobile versions during testing and so functions
// like .simulate('change') will not work, this workaround ensures desktop
// datepickers are used in tests instead
// https://github.com/mui/material-ui-pickers/issues/2073
export const applyDatePickerWorkaround = (): void => {
  // add window.matchMedia
  // this is necessary for the date picker to be rendered in desktop mode.
  // if this is not provided, the mobile mode is rendered, which might lead to unexpected behavior
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: (query: string) => ({
      media: query,
      // this is the media query that @material-ui/pickers uses to determine if a device is a desktop device
      matches: query === '(pointer: fine)',
      onchange: () => {},
      addEventListener: () => {},
      removeEventListener: () => {},
      addListener: () => {},
      removeListener: () => {},
      dispatchEvent: () => false,
    }),
  });
};

export const cleanupDatePickerWorkaround = (): void => {
  delete window.matchMedia;
};
