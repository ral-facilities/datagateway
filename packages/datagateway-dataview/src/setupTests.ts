/* eslint-disable @typescript-eslint/no-empty-function */
import '@testing-library/jest-dom';
// Blob implementation in jsdom is not complete (https://github.com/jsdom/jsdom/issues/2555)
// blob-polyfill fills in the gap
import 'blob-polyfill';
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { Action } from 'redux';
import { StateType } from './state/app.types';
import { initialState as dgDataViewInitialState } from './state/reducers/dgdataview.reducer';
import { dGCommonInitialState } from 'datagateway-common';

// Unofficial React 17 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() });

function noOp(): void {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
}

if (typeof window.URL.createObjectURL === 'undefined') {
  Object.defineProperty(window.URL, 'createObjectURL', { value: noOp });
}

// these are used for testing async actions
export let actions: Action[] = [];
export const resetActions = (): void => {
  actions = [];
};
export const getState = (): Partial<StateType> => ({
  dgdataview: dgDataViewInitialState,
  dgcommon: dGCommonInitialState,
});
export const dispatch = (action: Action): void | Promise<void> => {
  if (typeof action === 'function') {
    action(dispatch, getState);
    return Promise.resolve();
  } else {
    actions.push(action);
  }
};

export const flushPromises = (): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve));

// Mock lodash.debounce to return the function we want to call.
jest.mock('lodash.debounce', () => (fn: (args: unknown) => unknown) => fn);

// MUI date pickers default to mobile versions during testing and so functions
// like .simulate('change') will not work, this workaround ensures desktop
// datepickers are used in tests instead
// https://github.com/mui/material-ui-pickers/issues/2073
export const applyDatePickerWorkaround = (): void => {
  // add window.matchMedia
  // this is necessary for the date picker to be rendered in desktop mode.
  // if this is not provided, the mobile mode is rendered, which might lead to unexpected behavior
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: (query: string) => ({
      media: query,
      // this is the media query that @material-ui/pickers uses to determine if a device is a desktop device
      matches: query === '(pointer: fine)',
      onchange: () => {},
      addEventListener: () => {},
      removeEventListener: () => {},
      addListener: () => {},
      removeListener: () => {},
      dispatchEvent: () => false,
    }),
  });
};

export const cleanupDatePickerWorkaround = (): void => {
  delete window.matchMedia;
};
