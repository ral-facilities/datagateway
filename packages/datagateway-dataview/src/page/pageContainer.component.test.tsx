import React from 'react';
import { ReactWrapper, mount } from 'enzyme';

import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { StateType } from '../state/app.types';
import { initialState as dgDataViewInitialState } from '../state/reducers/dgdataview.reducer';
import {
  dGCommonInitialState,
  readSciGatewayToken,
  useCart,
  ClearFiltersButton,
} from 'datagateway-common';

import { LinearProgress } from '@mui/material';
import { createLocation, createMemoryHistory, History } from 'history';
import { Router } from 'react-router-dom';

import PageContainer, { paths } from './pageContainer.component';
import { checkInstrumentId, checkInvestigationId } from './idCheckFunctions';
import axios from 'axios';
import { act } from 'react-dom/test-utils';
import { flushPromises } from '../setupTests';
import {
  QueryClient,
  QueryClientProvider,
  useIsFetching,
  useQueryClient,
} from 'react-query';
import { Provider } from 'react-redux';

jest.mock('loglevel');
jest.mock('./idCheckFunctions');

jest.mock('datagateway-common', () => {
  const originalModule = jest.requireActual('datagateway-common');

  return {
    __esModule: true,
    ...originalModule,
    useCart: jest.fn(() => ({ data: [] })),
    // mock table and cardview to opt out of rendering them in these tests as there's no need
    Table: jest.fn(() => 'MockedTable'),
    CardView: jest.fn(() => 'MockedCardView'),
    readSciGatewayToken: jest.fn(() => originalModule.readSciGatewayToken()),
  };
});

jest.mock('react-query', () => ({
  __esModule: true,
  ...jest.requireActual('react-query'),
  useQueryClient: jest.fn(() => ({
    getQueryData: jest.fn(() => 0),
  })),
  useIsFetching: jest.fn(() => 0),
}));

describe('PageContainer - Tests', () => {
  let queryClient: QueryClient;
  let history: History;

  const createWrapper = (
    h: History = history,
    client: QueryClient = queryClient
  ): ReactWrapper => {
    const state: StateType = {
      dgcommon: dGCommonInitialState,
      dgdataview: dgDataViewInitialState,
      router: {
        action: 'POP',
        location: createLocation('/'),
      },
    };
    const mockStore = configureStore([thunk]);
    const testStore = mockStore(state);
    return mount(
      <Provider store={testStore}>
        <Router history={h}>
          <QueryClientProvider client={client}>
            <PageContainer />
          </QueryClientProvider>
        </Router>
      </Provider>
    );
  };

  beforeEach(() => {
    (axios.get as jest.Mock).mockImplementation((url: string) => {
      if (url.includes('count')) {
        return Promise.resolve({ data: 0 });
      } else {
        return Promise.resolve({ data: [] });
      }
    });
    queryClient = new QueryClient();
    history = createMemoryHistory({
      initialEntries: ['/'],
    });
    (useQueryClient as jest.Mock).mockReturnValue({
      getQueryData: jest.fn(() => 0),
    });
  });

  afterEach(() => {
    (useCart as jest.Mock).mockClear();
  });

  it('displays the correct entity count', () => {
    history.replace(paths.toggle.investigation);
    (useQueryClient as jest.Mock).mockReturnValue({
      getQueryData: jest.fn(() => 101),
    });

    const wrapper = createWrapper();

    expect(
      wrapper.find('[aria-label="view-count"]').first().find('h3').text()
    ).toBe('app.results: 101');
  });

  it('fetches cart on mount', () => {
    (useCart as jest.Mock).mockReturnValueOnce({
      data: [
        {
          entityId: 1,
          entityType: 'dataset',
          id: 1,
          name: 'Test 1',
          parentEntities: [],
        },
      ],
    });

    createWrapper();

    expect(useCart).toHaveBeenCalled();
  });

  it('opens search plugin when icon clicked', () => {
    const wrapper = createWrapper();

    wrapper.find('[aria-label="view-search"]').last().simulate('click');

    expect(history.location.pathname).toBe('/search/data');

    history.push('/browse/instrument');
    wrapper.find('[aria-label="view-search"]').last().simulate('click');

    expect(history.location.pathname).toBe('/search/isis');

    history.push('/browse/proposal');
    wrapper.find('[aria-label="view-search"]').last().simulate('click');

    expect(history.location.pathname).toBe('/search/dls');
  });

  it('opens download plugin when Download Cart clicked', () => {
    const wrapper = createWrapper();

    wrapper.find('[aria-label="app.cart_arialabel"]').last().simulate('click');

    expect(history.length).toBe(2);
    expect(history.location.pathname).toBe('/download');
  });

  it('do not display loading bar loading false', () => {
    const wrapper = createWrapper();

    expect(wrapper.exists(LinearProgress)).toBeFalsy();
  });

  it('display loading bar when loading true', () => {
    (useIsFetching as jest.Mock).mockReturnValueOnce(1);
    const wrapper = createWrapper();

    expect(wrapper.exists(LinearProgress)).toBeTruthy();
  });

  it('display clear filters button and clear for filters onClick', () => {
    history.replace(
      '/browse/investigation?filters=%7B"title"%3A%7B"value"%3A"spend"%2C"type"%3A"include"%7D%7D'
    );
    const wrapper = createWrapper();

    expect(wrapper.find(ClearFiltersButton).prop('disabled')).toEqual(false);

    wrapper
      .find('[data-testid="clear-filters-button"]')
      .last()
      .simulate('click');

    wrapper.update();

    expect(wrapper.find(ClearFiltersButton).prop('disabled')).toEqual(true);
    expect(history.location.search).toEqual('?');
  });

  it('display clear filters button and clear for filters onClick (/my-data/DLS)', () => {
    const dateNow = `${new Date(Date.now()).toISOString().split('T')[0]}`;
    history.replace(
      '/my-data/DLS?filters=%7B"startDate"%3A%7B"endDate"%3A" ' +
        dateNow +
        '"%7D%2C"title"%3A%7B"value"%3A"test"%2C"type"%3A"include"%7D%7D&sort=%7B"startDate"%3A"desc"%7D'
    );
    const response = { username: 'SomePerson' };
    (readSciGatewayToken as jest.Mock).mockReturnValue(response);
    const wrapper = createWrapper();

    expect(wrapper.find(ClearFiltersButton).prop('disabled')).toEqual(false);

    wrapper
      .find('[data-testid="clear-filters-button"]')
      .last()
      .simulate('click');

    wrapper.update();

    expect(wrapper.find(ClearFiltersButton).prop('disabled')).toEqual(true);

    expect(history.location.search).toEqual(
      '?filters=%7B%22startDate%22%3A%7B%22endDate%22%3A%22' +
        dateNow +
        '%22%7D%7D'
    );

    (readSciGatewayToken as jest.Mock).mockClear();
  });

  it('display disabled clear filters button', () => {
    history.replace(paths.toggle.investigation);
    const wrapper = createWrapper();

    expect(wrapper.find(ClearFiltersButton).prop('disabled')).toEqual(true);
  });

  it('display filter warning on datafile table', async () => {
    history.replace('/browse/investigation/1/dataset/25/datafile');
    (checkInvestigationId as jest.Mock).mockResolvedValueOnce(true);

    const wrapper = createWrapper();

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });
    expect(
      wrapper.find('[aria-label="filter-message"]').first().text()
    ).toEqual('loading.filter_message');
  });

  it('switches view button display name when clicked', () => {
    history.replace(paths.toggle.investigation);

    const wrapper = createWrapper();

    expect(
      wrapper.find('[aria-label="page view app.view_cards"]').exists()
    ).toBeTruthy();
    expect(
      wrapper.find('[aria-label="page view app.view_cards"]').first().text()
    ).toEqual('app.view_cards');

    // Click view button
    wrapper
      .find('[aria-label="page view app.view_cards"]')
      .last()
      .simulate('click');
    wrapper.update();

    // Check that the text on the button has changed
    expect(
      wrapper.find('[aria-label="page view app.view_table"]').first().text()
    ).toEqual('app.view_table');
  });

  it('displays role selector when on My Data route', () => {
    history.replace(paths.myData.root);

    const wrapper = createWrapper();

    expect(wrapper.find('#role-selector').exists()).toBeTruthy();
  });

  it('display filter warning on toggle table', () => {
    history.replace(`${paths.toggle.investigation}?view=table`);

    const wrapper = createWrapper();

    expect(
      wrapper.find('[aria-label="filter-message"]').first().text()
    ).toEqual('loading.filter_message');
  });

  it('do not display filter warning on toggle card', () => {
    history.replace(`${paths.toggle.investigation}?view=card`);

    const wrapper = createWrapper();

    expect(wrapper.exists('[aria-label="filter-message"]')).toBeFalsy();
  });

  it('do not use StyledRouting component on landing pages', async () => {
    (checkInstrumentId as jest.Mock).mockResolvedValueOnce(true);
    (useQueryClient as jest.Mock).mockReturnValue({
      getQueryData: jest.fn(),
    });
    history.replace(`${paths.studyHierarchy.landing.isisStudyLanding}`);

    const wrapper = createWrapper();

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    expect(wrapper.exists('StyledRouting')).toBeFalsy();
  });

  it('set view to card if cardview stored in localstorage', () => {
    localStorage.setItem('dataView', 'card');
    history.replace(paths.toggle.investigation);

    createWrapper();

    expect(history.location.search).toBe('?view=card');
  });

  it('displays warning label when browsing data anonymously', () => {
    const response = { username: 'anon/anon' };
    (readSciGatewayToken as jest.Mock).mockReturnValueOnce(response);

    const wrapper = createWrapper();

    expect(
      wrapper.find('[aria-label="open-data-warning"]').exists()
    ).toBeTruthy();
  });

  it('displays warning label when browsing study hierarchy', () => {
    history.replace(paths.studyHierarchy.toggle.isisStudy);
    const response = { username: 'SomePerson' };
    (readSciGatewayToken as jest.Mock).mockReturnValueOnce(response);

    const wrapper = createWrapper();

    expect(
      wrapper.find('[aria-label="open-data-warning"]').exists()
    ).toBeTruthy();
  });

  it('does not display warning label when logged in', () => {
    const response = { username: 'SomePerson' };
    (readSciGatewayToken as jest.Mock).mockReturnValueOnce(response);

    const wrapper = createWrapper();

    expect(
      wrapper.find('[aria-label="open-data-warning"]').exists()
    ).toBeFalsy();
  });

  it('shows SelectionAlert banner when item selected', () => {
    // Supply data to make SelectionAlert display
    (useCart as jest.Mock).mockReturnValueOnce({
      data: [
        {
          entityId: 1,
          entityType: 'dataset',
          id: 1,
          name: 'Test 1',
          parentEntities: [],
        },
      ],
    });
    const wrapper = createWrapper();

    expect(wrapper.exists('[aria-label="selection-alert"]')).toBeTruthy();
  });

  it('does not show SelectionAlert banner when no items are selected', () => {
    (useCart as jest.Mock).mockReturnValueOnce({
      data: [],
    });
    const wrapper = createWrapper();

    expect(wrapper.exists('[aria-label="selection-alert"]')).toBeFalsy();
  });

  it('opens download plugin when link in SelectionAlert clicked', () => {
    // Supply data to make SelectionAlert display
    (useCart as jest.Mock).mockReturnValueOnce({
      data: [
        {
          entityId: 1,
          entityType: 'dataset',
          id: 1,
          name: 'Test 1',
          parentEntities: [],
        },
      ],
    });
    const wrapper = createWrapper();

    wrapper
      .find('[aria-label="selection-alert-link"]')
      .first()
      .simulate('click');

    expect(history.location.pathname).toBe('/download');
  });

  it('passes correct landing page entities to breadcrumbs', () => {
    history.replace(paths.toggle.isisInvestigation);
    let wrapper = createWrapper();

    expect(wrapper.find('PageBreadcrumbs').prop('landingPageEntities')).toEqual(
      ['investigation', 'dataset']
    );

    history.replace(paths.studyHierarchy.toggle.isisInvestigation);
    wrapper = createWrapper();

    expect(wrapper.find('PageBreadcrumbs').prop('landingPageEntities')).toEqual(
      ['study', 'investigation', 'dataset']
    );
  });

  it('does not fetch cart when on homepage (cart request errors when user is viewing homepage unauthenticated)', () => {
    history.replace(paths.homepage);
    createWrapper();

    expect(useCart).not.toHaveBeenCalled();
  });
});
