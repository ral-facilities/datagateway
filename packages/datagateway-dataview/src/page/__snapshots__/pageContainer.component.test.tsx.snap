// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageContainer - Tests displays the correct entity count 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation0"
  style={
    Object {
      "backgroundColor": "inherit",
    }
  }
>
  <NavBar
    cartItems={Array []}
    entityCount={101}
    navigateToDownload={[Function]}
    navigateToSearch={[Function]}
  />
  <WithStyles(WithStyles(ForwardRef(Grid)))
    container={true}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      xs={12}
    >
      <Route
        exact={true}
        path={
          Array [
            "/browse/investigation",
            "/browse/investigation/:investigationId/dataset",
            "/browse/instrument",
            "/browse/instrument/:instrumentId/facilityCycle",
            "/browse/instrument/:instrumentId/facilityCycle/:facilityCycleId/investigation",
            "/browse/instrument/:instrumentId/facilityCycle/:facilityCycleId/investigation/:investigationId/dataset",
            "/browse/proposal",
            "/browse/proposal/:proposalName/investigation",
            "/browse/proposal/:proposalName/investigation/:investigationId/dataset",
            "/browseStudyHierarchy/instrument",
            "/browseStudyHierarchy/instrument/:instrumentId/study",
            "/browseStudyHierarchy/instrument/:instrumentId/study/:studyId/investigation",
            "/browseStudyHierarchy/instrument/:instrumentId/study/:studyId/investigation/:investigationId/dataset",
          ]
        }
        render={[Function]}
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      aria-label="container-view"
      item={true}
      xs={12}
    />
  </WithStyles(WithStyles(ForwardRef(Grid)))>
</div>
`;
